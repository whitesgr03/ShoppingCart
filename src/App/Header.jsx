import { useState, useEffect } from "react";
import { NavLink, useNavigate } from "react-router-dom";
import { initialAuth } from "../firebase-config";

import { useCart, useCartDispatch, useModalDispatch } from "./RootContext";

import {
	userLogout,
	googleLogin,
	checkUser,
	createUser,
} from "../utils/handleUserAccount";
import { getUserCart } from "../utils/handleUserCarts";

import Icon from "@mdi/react";
import { mdiCartVariant, mdiAccount, mdiTextBoxOutline } from "@mdi/js";

import { useCart, useModalDispatch } from "./RootContext";

const HeaderBadge = () => {
	const cartList = useCart();

	return (
		<>
			{cartList.length > 0 && (
				<span data-testid="badge" className="badge">
					{cartList.length}
				</span>
			)}
		</>
	);
};

const Header = () => {
	const [isLogin, setIsLogin] = useState(false);
	const [auth, setAuth] = useState(true);
	const navigate = useNavigate();

	const modalDispatch = useModalDispatch();
	const cartDispatch = useCartDispatch();

	const handlePointerUp = () => {
		modalDispatch({
			type: "cart",
		});
	};

	useEffect(() => {
		const auth = initialAuth();

		const unsubscribe = auth.onAuthStateChanged(async user => {
			const hasUser = user && (await checkUser(user.uid));

			if (hasUser && !hasUser.success) {
				console.error(hasUser.message);
				return;
			}

			const createNewUser =
				user && !hasUser.data.userExists && (await createUser(user));

			if (createNewUser && !createNewUser.success) {
				console.error(hasUser.message);
				return;
			}

			const getCart = user && (await getUserCart());

			if (getCart && !getCart.success) {
				console.error(getCart.message);
				navigate("/error");
				return;
			}

			cartDispatch({
				type: "initialize_cart",
				cart: user && getCart.success ? getCart.data : [],
			});

			setIsLogin(user ? true : false);
			setAuth(false);
		});

		return () => unsubscribe();
	}, [cartDispatch, navigate]);

	return (
		<div className="sidebar" data-testid="sidebar">
			<div className="icons">
				<button
					className="account button"
					type="button"
					name="showAccount"
				>
					<Icon path={mdiAccount} />
				</button>
				<button className="order button" type="button" name="showOrder">
					<Icon path={mdiTextBoxOutline} />
				</button>
				<button
					className="cart button"
					type="button"
					name="showCart"
					data-testid="cart"
					onPointerUp={handlePointerUp}
				>
					<Icon path={mdiCartVariant} />
					<HeaderBadge />
				</button>
			</div>
			<h1 className="title">
				<NavLink to="/">GentSkin</NavLink>
			</h1>
			<ul className="pages">
				<li>
					<NavLink to="/">Home</NavLink>
				</li>
				<li>
					<NavLink to="shop">Shop</NavLink>
				</li>
				<li>
					<NavLink to="contact">Contact</NavLink>
				</li>
			</ul>
		</div>
	);
};

export { Header as default, HeaderBadge };
