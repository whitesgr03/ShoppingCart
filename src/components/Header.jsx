import Icon from "@mdi/react";
import { mdiCartVariant, mdiAccount, mdiTextBoxOutline } from "@mdi/js";

import { NavLink } from "react-router-dom";

const Header = ({ cartList, onToggleModal }) => {
	const handleOpenModal = () => onToggleModal(true);
	return (
		<div className="sidebar">
			<div className="icons">
				<button type="button">
					<Icon path={mdiAccount} />
				</button>
				<button type="button">
					<Icon path={mdiTextBoxOutline} />
				</button>
				<button type="button">
					<Icon path={mdiCartVariant} />
				</button>
			</div>
			<h1 className="title">
				<NavLink to="/">GentSkin</NavLink>
			</h1>
			<ul className="pages">
				<li>
					<NavLink to="/">Home</NavLink>
				</li>
				<li>
					<NavLink to="shop">Shop</NavLink>
				</li>
				<li>
					<NavLink to="contact">Contact</NavLink>
				</li>
			</ul>
		</div>
	);
};

export { Header as default };
