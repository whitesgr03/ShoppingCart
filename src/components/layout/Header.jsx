import "../../style/layout/header.css";
import Icon from "@mdi/react";
import {
	mdiCartVariant,
	mdiAccount,
	mdiTextBoxOutline,
	mdiLogout,
} from "@mdi/js";

import { NavLink } from "react-router-dom";
import {
	handleGoogleLogin,
	handleUserLogout,
} from "../../utils/handleUserAccount";

import PropTypes from "prop-types";

const Badge = ({ cart }) => {
	return (
		cart.length > 0 && (
			<span data-testid="badge" className="badge">
				{cart.length}
			</span>
		)
	);
};

const Header = ({ userId, setAppError, onOpenModal, children }) => {
	const handleLogin = async () => {
		try {
			await handleGoogleLogin();
		} catch (error) {
			setAppError("Service temporarily unavailable");
		}
	};

	const handleLogout = async () => {
		try {
			await handleUserLogout();
		} catch (error) {
			setAppError("Service temporarily unavailable");
		}
	};

	return (
		<div className="sidebar">
			<div
				data-testid="background"
				className={`icons ${userId === null ? "authenticate" : ""}`}
			>
				<button
					className="account"
					type="button"
					name="showAccount"
					data-testid="accountButton"
					onClick={userId === "" ? handleLogin : handleLogout}
				>
					<Icon path={userId === "" ? mdiAccount : mdiLogout} />
				</button>
				<button className="order" type="button" name="showOrder">
					<Icon path={mdiTextBoxOutline} />
				</button>
				<button
					className="cart"
					type="button"
					name="showCart"
					data-testid="cartButton"
					onClick={() => onOpenModal("cart")}
				>
					<Icon path={mdiCartVariant} />
					{children}
				</button>
			</div>
			<h1 className="title">
				<NavLink to="/">GentSkin</NavLink>
			</h1>
			<ul className="pages">
				<li>
					<NavLink to="/">Home</NavLink>
				</li>
				<li>
					<NavLink to="shop">Shop</NavLink>
				</li>
				<li>
					<NavLink to="contact">Contact</NavLink>
				</li>
			</ul>
		</div>
	);
};

Header.propTypes = {
	userId: PropTypes.string,
	setAppError: PropTypes.func,
	onOpenModal: PropTypes.func,
	children: PropTypes.node.isRequired,
};

Badge.propTypes = {
	isLoading: PropTypes.array,
};

export { Header as default, Badge };
