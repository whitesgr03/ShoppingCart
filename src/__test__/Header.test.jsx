import { render } from "@testing-library/react";

import { MemoryRouter } from "react-router-dom";

import Header from "../components/Header";

describe("Renders Header Component", () => {
	it("Should return Header", () => {
		const { container } = render(
			<MemoryRouter initialEntries={["/"]}>
				<Header />
			</MemoryRouter>
		);

		const actual = container;

		expect(actual).toMatchSnapshot();
	});
});
