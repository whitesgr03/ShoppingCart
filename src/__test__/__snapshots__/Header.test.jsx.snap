// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Header Component Should return Header DOM with empty cartList 1`] = `
<div>
  <div
    class="sidebar"
  >
    <div
      class="icons"
    >
      <button
        type="button"
      >
        <svg
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12,4A4,4 0 0,1 16,8A4,4 0 0,1 12,12A4,4 0 0,1 8,8A4,4 0 0,1 12,4M12,14C16.42,14 20,15.79 20,18V20H4V18C4,15.79 7.58,14 12,14Z"
            style="fill: currentColor;"
          />
        </svg>
      </button>
      <button
        type="button"
      >
        <svg
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M5,3C3.89,3 3,3.89 3,5V19C3,20.11 3.89,21 5,21H19C20.11,21 21,20.11 21,19V5C21,3.89 20.11,3 19,3H5M5,5H19V19H5V5M7,7V9H17V7H7M7,11V13H17V11H7M7,15V17H14V15H7Z"
            style="fill: currentColor;"
          />
        </svg>
      </button>
      <button
        class="cart button"
        name="showCart"
        type="button"
      >
        <svg
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 20C19 21.11 18.11 22 17 22C15.89 22 15 21.1 15 20C15 18.89 15.89 18 17 18C18.11 18 19 18.9 19 20M7 18C5.89 18 5 18.89 5 20C5 21.1 5.89 22 7 22C8.11 22 9 21.11 9 20S8.11 18 7 18M7.2 14.63L7.17 14.75C7.17 14.89 7.28 15 7.42 15H19V17H7C5.89 17 5 16.1 5 15C5 14.65 5.09 14.32 5.24 14.04L6.6 11.59L3 4H1V2H4.27L5.21 4H20C20.55 4 21 4.45 21 5C21 5.17 20.95 5.34 20.88 5.5L17.3 11.97C16.96 12.58 16.3 13 15.55 13H8.1L7.2 14.63M8.5 11H10V9H7.56L8.5 11M11 9V11H14V9H11M14 8V6H11V8H14M17.11 9H15V11H16L17.11 9M18.78 6H15V8H17.67L18.78 6M6.14 6L7.08 8H10V6H6.14Z"
            style="fill: currentColor;"
          />
        </svg>
      </button>
    </div>
    <h1
      class="title"
    >
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        GentSkin
      </a>
    </h1>
    <ul
      class="pages"
    >
      <li>
        <a
          aria-current="page"
          class="active"
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          class=""
          href="/shop"
        >
          Shop
        </a>
      </li>
      <li>
        <a
          class=""
          href="/contact"
        >
          Contact
        </a>
      </li>
    </ul>
  </div>
</div>
`;
